logstash-core/src/main/java/org/logstash/config/ir/compiler/Dataset.java

```diff
@@ -0,0 +1,438 @@
+
+            @Override
+            public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+                final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+                final RubyHash options) {
+                return EMPTY_RETURN;
+            }
+
+            @Override
+            public void clear() {
+            }
+        };
+
+        private final Collection<Dataset> parents;
+
+        /**
+         * <p>Builds a terminal {@link Dataset} from the given parent {@link Dataset}s.</p>
+         * <p>If the given set of parent {@link Dataset} is empty the sum is defined as the
+         * trivial dataset that does not invoke any computation whatsoever.</p>
+         * The return of a call to
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)} is always
+         * {@link Collections#emptyList()}.
+         * @param parents Parent {@link Dataset} to sum and terminate
+         * @return Dataset representing the sum of given parent {@link Dataset}
+         */
+        public static Dataset from(final Collection<Dataset> parents, final boolean debug) {
+            final int count = parents.size();
+            final Dataset result;
+            if (count > 0) {
+                if (debug) {
+                    result = new Dataset.TerminalDataset.TerminalDebugDataset(parents);
+                } else {
+                    result = new Dataset.TerminalDataset(parents);
+                }
+            } else {
+                result = TRIVIAL;
+            }
+            return result;
+        }
+
+        private TerminalDataset(final Collection<Dataset> parents) {
+            this.parents = parents;
+        }
+
+        @Override
+        public Collection<JrubyEventExtLibrary.RubyEvent> compute(
+            final RubyIntegration.Batch batch, final boolean flush, final boolean shutdown,
+            final RubyHash options) {
+            parents.forEach(dataset -> dataset.compute(batch, flush, shutdown, options));
+            this.clear();
+            return EMPTY_RETURN;
+        }
+
+        @Override
+        public void clear() {
+            for (final Dataset parent : parents) {
+                parent.clear();
+            }
+        }
+
+        /**
+         * <p>{@link Dataset} that contains all {@link JrubyEventExtLibrary.RubyEvent} of all of
+         * its dependencies and is assumed to be at the end of an execution.</p>
+         * This dataset does not require an explicit call to {@link Dataset#clear()} since it will
+         * automatically {@code clear} all of its parents.
+         * Unlike {@link Dataset.TerminalDataset} this class will return all
+         * {@link JrubyEventExtLibrary.RubyEvent} that passed through it on a call to
+         * {@link Dataset#compute(RubyIntegration.Batch, boolean, boolean, RubyHash)}.
+         */
+        public static final class TerminalDebugDataset implements Dataset {
```

Comment:
@jordansissel the `TerminalDataset` is a direct child of the `interface` `Dataset` => it's automatically `public` without the `public` keyword :)